File: java/Longest Substring Without Repeating Characters.java
// Given a string, find the length of the longest substring without repeating characters. 
// For example, the longest substring without repeating letters for "abcabcbb" is "abc", 
// which the length is 3. For "bbbbb" the longest substring is "b", with the length of 1.

public class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.length()==0){
            return 0;
        }
        
        Map<Character,Integer> ha=new HashMap<Character,Integer>();
        int length=0;
        int max=0;
        
        for(int i =0;i<s.length();i++){
            length++;
            if(ha.containsKey(s.charAt(i))){
                length=Math.min(length,i-ha.get(s.charAt(i)));
            }
            ha.put(s.charAt(i),i);
            max=Math.max(length,max);
        }
        return max;
    }
}

File: c++/Longest Substring Without Repeating Characters.cpp
/*Given a string, find the length of the longest substring without repeating characters. For example, the longest substring without repeating letters for "abcabcbb" is "abc", which the length is 3. For "bbbbb" the longest substring is "b", with the length of 1.
*/
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
    	unordered_map<char,int> ump;
    	int res = 0;
    	int start = 0;
    	for(string::size_type i=0;i<s.size();i++){
    		if(ump.find(s[i])!=ump.end()){
    			res = max(res,(int)i-start);
    			start = max(ump[s[i]]+1,start);
    		}
    		ump[s[i]] = i;
    	}
    	res = max(res,(int)s.size()-start);
    	return res;
    }
};


